processing/src/main/java/io/druid/query/search/search/AutoStrategy.java

```diff
@@ -0,0 +1,60 @@
+package io.druid.query.search.search;
+
+import com.metamx.emitter.EmittingLogger;
+import io.druid.collections.bitmap.ImmutableBitmap;
+import io.druid.java.util.common.IAE;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.filter.Filter;
+import io.druid.query.search.search.CursorBasedStrategy.CursorBasedExecutor;
+import io.druid.query.search.search.IndexOnlyStrategy.IndexOnlyExecutor;
+import io.druid.segment.QueryableIndex;
+import io.druid.segment.Segment;
+import io.druid.segment.filter.Filters;
+import org.joda.time.Interval;
+
+import java.util.List;
+
+public class AutoStrategy extends SearchStrategy
+{
+  private static final EmittingLogger log = new EmittingLogger(AutoStrategy.class);
+
+  @Override
+  public SearchQueryExecutor getExecutionPlan(SearchQuery query, Segment segment)
+  {
+    final QueryableIndex index = segment.asQueryableIndex();
+
+    if (index != null) {
+      final Filter filter = Filters.convertToCNFFromQueryContext(query, Filters.toFilter(query.getDimensionsFilter()));
+      final List<Interval> intervals = query.getQuerySegmentSpec().getIntervals();
+      if (intervals.size() != 1) {
+        throw new IAE("Should only have one interval, got[%s]", intervals);
+      }
+      final Interval interval = intervals.get(0);
+      final ImmutableBitmap timeFilteredBitmap = IndexOnlyExecutor.makeTimeFilteredBitmap(index,
```

Comment:
There's a couple of performance issues here:

1. `makeTimeFilteredBitmap` isn't cheap, and if we use the index-based execution we will end up calling it twice.
2. `filter.getBitmapIndex` isn't cheap either, and if we end up doing the cursor-based execution, it'll get called twice too (once by makeTimeFilteredBitmap and once by makeCursors).

Some possible strategies to help:

- Don't generate the timeFilteredBitmap until we actually need to call `hasHighSelectivity`. i.e. if the search dims are low cardinality, just go straight to the index-only strategy. And if they're very high cardinality, similar to the number of rows in the segment, then possibly go straight to the cursor-only strategy (please do some benchmarks to verify this guess though). We should only need to check the filter selectivity if search dimension cardinality is medium.
- Save the timeFilteredBitmap after generating it, and pass it to the index-based executor if we choose that one.

I think one thing we should _not_ do is save the bitmap and pass it to makeCursors. That's probably too leaky and I would rather accept the performance hit there.